### ISSUE TITLE
ValueTuple compatibility issues in .NET Framework 4.7.1

### ISSUE THREAD

**Issue by mkosieradzki** [dotnet/corefx#24653]
Cross posting for better visibility:
Microsoft/dotnet-framework-early-access#9

Fall creators update is going live in two days. Heads up! It will generate nasty compat issues for applications using `Task<ValueTuple>`. Not sure if this is a known issue or by design, but I have lost entire day today trying to track it down.

**Comment by karelz** [comment]
@ericstj @weshaggard @jcouv is this a known problem?

**Comment by weshaggard** [comment]
@AlexGhiondea is this problem known? If not we probably should investigate to understand it.

**Comment by AlexGhiondea** [comment]
This was not know. But what is happening is that the System.ValueTuple that gets loaded is the one carried with the app -- there is a binding redirect in the .config file that forces that. Removing the binding redirect allows the app to work.

**Comment by mkosieradzki** [comment]
Yup. Commenting out this single binding redirect:
```
<assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" />
      </dependentAssembly>
    </assemblyBinding>
```
Makes the application run without any issues.

**Comment by weshaggard** [comment]
@AlexGhiondea so binding redirects overwrite the unification table? If that is the case we might be hitting other similar issues with other things we moved inbox.

**Comment by mkosieradzki** [comment]
If I understand correctly .NET Framework 4.7.1 today starts getting rolled out to customers through the Windows Update (as the Fall Creators Update)? It's Oct 17th... I am a bit afraid it might break some apps... including ours... before you even have a chance to fix this issue.

**Comment by AlexGhiondea** [comment]
@weshaggard yes, the binding redirects override the unification table.

**Comment by marchy** [comment]
After the Windows Update our code now fails at runtime with "Method not found: 'System.ValueTuple`2<System.String,System.String>" wherever tuples get invoked.

Cannot seem to find a way around this. Our codebase targets 'net462' (as to run on Azure) but local machines are now on 4.7.1. What to do?

**Comment by mkosieradzki** [comment]
@marchy The only reasonable workaround I know is to retarget to net47.

**Comment by AlexGhiondea** [comment]
Hey @marchy this is a known issue in 4.7.1. More information about it is [here](https://github.com/Microsoft/dotnet/blob/master/releases/net471/KnownIssues/510901-BCL%20Apps%20targeting%20.NET-4.6.1%20that%20use%20.NET%20Standard%20libraries%20might%20be%20broken%20when%20running%20on%20.NET%204.7.1.md)

Currently the workaround is to either manually remove the binding redirects to System.ValueTuple of, as @mkosieradzki mentioned, re-target the application to 4.7 or 4.7.1.

We are working on a fix for this.

**Comment by marchy** [comment]
Thanks for the very prompt reply @AlexGhiondea  @mkosieradzki.

Re-targeting to 4.7 solved it :)
(while re-targeting to 4.7.1 is not feasible since apparently Azure doesn't expect support for it until December)

**Comment by joperezr** [comment]
Closing this issue as this problem is now being tracked in a different repo: https://github.com/dotnet/standard/issues/567

Just as a small FYI, this problem has been addressed in 4.7.2 and a fix for it will be serviced to 4.7.1 machines.
